modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteIndexCount.java

```diff
@@ -0,0 +1,128 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.calcite.rel;
+
+import java.util.List;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptCost;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelOptTable;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.AbstractRelNode;
+import org.apache.calcite.rel.RelInput;
+import org.apache.calcite.rel.RelWriter;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rel.type.RelDataType;
+
+import static org.apache.ignite.internal.processors.query.calcite.metadata.cost.IgniteCost.ROW_PASS_THROUGH_COST;
+
+/**
+ * Returns number of index records.
+ */
+public class IgniteIndexCount extends AbstractRelNode implements SourceAwareIgniteRel {
+    /** */
+    private final RelOptTable tbl;
+
+    /** */
+    private final String idxName;
+
+    /**
+     * Constructor for deserialization.
+     *
+     * @param input Serialized representation.
+     */
+    public IgniteIndexCount(RelInput input) {
+        super(input.getCluster(), input.getTraitSet());
+
+        idxName = input.getString("index");
+        tbl = input.getTable("table");
+        rowType = input.getRowType("type");
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param cluster Cluster that this relational expression belongs to
+     * @param traits Traits of this relational expression
+     * @param tbl Table definition.
+     * @param idxName Index name.
+     * @param type Data type.
+     */
+    public IgniteIndexCount(
+        RelOptCluster cluster,
+        RelTraitSet traits,
+        RelOptTable tbl,
+        String idxName,
+        RelDataType type
+    ) {
+        super(cluster, traits);
+
+        this.idxName = idxName;
+        this.tbl = tbl;
+        this.rowType = type;
+    }
+
+    /** */
+    public String indexName() {
+        return idxName;
+    }
+
+    /** {@inheritDoc} */
+    @Override public double estimateRowCount(RelMetadataQuery mq) {
+        // Requesting index count always produces just one record.
+        return 1.0d;
+    }
+
+    /** {@inheritDoc} */
+    @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+        return planner.getCostFactory().makeCost(1.0, 1.0, tbl.getRowCount() * ROW_PASS_THROUGH_COST);
```

Comment:
But reading page involves the IO. It could bring page from disk, not from mem. Actually, we depends here on cluster size and backups. I would say the cost is something like 'tableSize / 1000' for REPLICATED or 'tableSize * (1 + backups) / clusterSize / 1000'. OK, lets keep it approximated tableSize / 1000.

Reply:
If we indroduce IO cost here than we should also calculate IO cost for other types of scan. I think it doesn't give us high value now, perhaps can be done in future, by another ticket.